File: subversion/bindings/java/javahl/src/org/tigris/subversion/javahl/Notify.java
/**
 * @copyright
 * ====================================================================
 * Copyright (c) 2003 CollabNet.  All rights reserved.
 *
 * This software is licensed as described in the file COPYING, which
 * you should have received as part of this distribution.  The terms
 * are also available at http://subversion.tigris.org/license-1.html.
 * If newer versions of this license are posted there, you may use a
 * newer version instead, at your option.
 *
 * This software consists of voluntary contributions made by many
 * individuals.  For exact contribution history, see the revision
 * history and logs, available at http://subversion.tigris.org/.
 * ====================================================================
 * @endcopyright
 */
package org.tigris.subversion.javahl;

/**
 * Subversion notification interface.
 *
 * Implement this interface and implement the onNotify method
 * to provide a custom notification handler to the Modify
 * class.
 * If you need to pass extra information to the notification
 * handler then just add these to you implementing class
 */
public interface Notify
{
    /**
     * Handler for Subversion notifications.
     *
     * Override this function to allow Subversion to
     * send notifications
     * @param path on which action happen
     * @param action subversion action, see svn_wc_notify_action_t
     * @param kind node kind of path after action occurred
     * @param mimeType mime type of path after action occurred
     * @param contentState state of content after action occurred
     * @param propState state of properties after action occurred
     * @param revision revision number  after action occurred
     */
    public void onNotify(String path, int action, int kind, String mimeType, int contentState, int propState, long revision);

    /** The type of action occuring. */
    public static final class Action
    {
        public static final int add = 0;
        public static final int copy = 1;
        public static final int delete =2;
        public static final int restore = 3;
        public static final int undo = 4;
        public static final int failed_undo = 5;
        public static final int resolve = 6;
        public static final int status = 7;
        public static final int skip = 8;

        /* The update actions are also used for checkouts, switches, and merges. */

        /** Got a delete in an update. */
        public static final int update_delete = 9;

        /** Got an add in an update. */
        public static final int update_add = 10;

        /** Got any other action in an update. */
        public static final int update_update = 11;

        /** The last notification in an update */
        public static final int update_completed = 12;

        /** About to update an external module, use for checkouts and switches too,
         * end with @c svn_wc_update_completed.
         */
        public static final int update_external = 13;

        public static final int commit_modified = 14;
        public static final int commit_added = 15;
        public static final int commit_deleted = 16;
        public static final int commit_replaced = 17;
        public static final int commit_postfix_txdelta = 18;
		private static final String[] actionNames =
		{
			"add",
			"copy",
			"delete",
			"restore",
			"undo",
			"failed undo",
			"resolve",
			"status",
			"skip",
			"update delete",
			"update add",
			"update modified",
			"update completed",
			"update external",
			"sending modified",
			"sending added   ",
			"sending deleted ",
			"sending replaced",
			"transfer"

		};
		public static final String getActionName(int action)
		{
			return actionNames[action];
		}

    }
    /** The type of notification that is occuring. */
    public static final class Status
    {
        public static final int inapplicable = 0;

        /** Notifier doesn't know or isn't saying. */
        public static final int unknown = 1;

        /** The state did not change. */
        public static final int unchanged = 2;

        /** Pristine state was modified. */
        public static final int changed = 3;

        /** Modified state had mods merged in. */
        public static final int merged = 4;

        /** Modified state got conflicting mods. */
        public static final int conflicted = 5;

		private static final String[] statusNames =
		{
			"inapplicable",
			"unknown",
			"unchanged",
			"changed",
			"merged",
			"conflicted",
		};
		public static final String getStatusName(int status)
		{
			return statusNames[status];
		}
    }

}

File: subversion/bindings/java/javahl/native/Notify.cpp
/**
 * @copyright
 * ====================================================================
 * Copyright (c) 2003 CollabNet.  All rights reserved.
 *
 * This software is licensed as described in the file COPYING, which
 * you should have received as part of this distribution.  The terms
 * are also available at http://subversion.tigris.org/license-1.html.
 * If newer versions of this license are posted there, you may use a
 * newer version instead, at your option.
 *
 * This software consists of voluntary contributions made by many
 * individuals.  For exact contribution history, see the revision
 * history and logs, available at http://subversion.tigris.org/.
 * ====================================================================
 * @endcopyright
 *
 * @file Notify.cpp
 * @brief Implementation of the class Notify
 */

#include "Notify.h"
#include "JNIUtil.h"
#include "EnumMapper.h"
//////////////////////////////////////////////////////////////////////
// Construction/Destruction
//////////////////////////////////////////////////////////////////////
/**
 * Create a new object and store the java object
 * @param notify    global reference to the java object
 */
Notify::Notify(jobject p_notify)
{
    m_notify = p_notify;
}

/**
 * Destroy the object and delete the global reference to the java object
 */ 
Notify::~Notify()
{
    if(m_notify != NULL)
    {
        JNIEnv *env = JNIUtil::getEnv();
        env->DeleteGlobalRef(m_notify);
    }
}

/**
 * Create a C++ peer object for the java object
 * @param notify    a local reference to the java object
 */
Notify * Notify::makeCNotify(jobject notify)
{
    // if the java object is null -> no C++ peer needed
    if(notify == NULL)
        return NULL;
    JNIEnv *env = JNIUtil::getEnv();

    // sanity check, that the object implements Notify
    jclass clazz = env->FindClass(JAVA_PACKAGE"/Notify");
    if(JNIUtil::isJavaExceptionThrown())
    {
        return NULL;
    }
    if(!env->IsInstanceOf(notify, clazz))
    {
        env->DeleteLocalRef(clazz);
        return NULL;
    }
    env->DeleteLocalRef(clazz);
    if(JNIUtil::isJavaExceptionThrown())
    {
        return NULL;
    }

    // make a global reference, because the reference is longer needed, than
    // the call
    jobject myNotify = env->NewGlobalRef(notify);
    if(JNIUtil::isJavaExceptionThrown())
    {
        return NULL;
    }

    // create the peer
    return new Notify(myNotify);
}
  /**
   * notification function passed as svn_wc_notify_func_t
   * @param baton notification instance is passed using this parameter
   * @param path on which action happen
   * @param action subversion action, see svn_wc_notify_action_t
   * @param kind node kind of path after action occurred
   * @param mime_type mime type of path after action occurred
   * @param content_state state of content after action occurred
   * @param prop_state state of properties after action occurred
   * @param revision revision number after action occurred
   */

void
Notify::notify (
    void *baton,
    const char *path,
    svn_wc_notify_action_t action,
    svn_node_kind_t kind,
    const char *mime_type,
    svn_wc_notify_state_t content_state,
    svn_wc_notify_state_t prop_state,
    svn_revnum_t revision)
{
    // an Notify object is used as the baton
    Notify * notify = (Notify *) baton;
    if(notify) // sanity check
    {
        // call our method
        notify->onNotify(path, action, kind, mime_type,
            content_state, prop_state, revision);
    }
}
  /**
   * Handler for Subversion notifications.
   *
   * @param path on which action happen
   * @param action subversion action, see svn_wc_notify_action_t
   * @param kind node kind of path after action occurred
   * @param mime_type mime type of path after action occurred
   * @param content_state state of content after action occurred
   * @param prop_state state of properties after action occurred
   * @param revision revision number  after action occurred
   */
void
Notify::onNotify (
    const char *path,
    svn_wc_notify_action_t action,
    svn_node_kind_t kind,
    const char *mime_type,
    svn_wc_notify_state_t content_state,
    svn_wc_notify_state_t prop_state,
    svn_revnum_t revision)
{
    JNIEnv *env = JNIUtil::getEnv();
    // java method id will not change during the time this library is loaded, 
    // so it can be cached. 
    static jmethodID mid = 0;
    if(mid == 0)
    {
        jclass clazz = env->FindClass(JAVA_PACKAGE"/Notify");
        //jclass clazz = env->GetObjectClass(m_notify);
        if(JNIUtil::isJavaExceptionThrown())
        {
            return;
        }
        mid = env->GetMethodID(clazz, "onNotify", "(Ljava/lang/String;IILjava/lang/String;IIJ)V");
        if(JNIUtil::isJavaExceptionThrown() || mid == 0)
        {
            return;
        }
        env->DeleteLocalRef(clazz);
        if(JNIUtil::isJavaExceptionThrown())
        {
            return;
        }
    }

    // convert the parameter to their java relatives
    jstring jPath = JNIUtil::makeJString(path);
    if(JNIUtil::isJavaExceptionThrown())
    {
        return;
    }

    jint jAction = EnumMapper::mapNotifyAction(action);
    jint jKind = EnumMapper::mapNodeKind(kind);
    jstring jMimeType = JNIUtil::makeJString(mime_type);
    if(JNIUtil::isJavaExceptionThrown())
    {
        return;
    }
    jint jContentState = EnumMapper::mapNotifyState(content_state);
    jint jPropState = EnumMapper::mapNotifyState(prop_state);

    // call the java method
    env->CallVoidMethod(m_notify, mid, jPath, jAction, jKind, jMimeType, jContentState, jPropState,
        (jlong)revision);
    if(JNIUtil::isJavaExceptionThrown())
    {
        return;
    }

    // release all the temporary java objects
    env->DeleteLocalRef(jPath);
    if(JNIUtil::isJavaExceptionThrown())
    {
        return;
    }
    env->DeleteLocalRef(jMimeType);
    if(JNIUtil::isJavaExceptionThrown())
    {
        return;
    }
}
